Zend/zend_alloc.c

```diff
@@ -3092,199 +3082,111 @@ ZEND_API void zend_mm_get_custom_handlers(zend_mm_heap *heap,
  */
 ZEND_API zend_mm_observer* zend_mm_observer_register(
+	zend_mm_heap* heap,
 	void (*malloc)(size_t, void * ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC),
 	void (*free)(void * ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC),
 	void (*realloc)(void *, size_t, void * ZEND_FILE_LINE_DC ZEND_FILE_LINE_ORIG_DC)
 ) {
 #if ZEND_MM_CUSTOM
+	if (!heap && !(heap = AG(mm_heap))) {
+		return false;
+	}
+
 	zend_mm_observer *node = pemalloc(sizeof(zend_mm_observer), 1);
 	node->malloc = malloc;
 	node->free = free;
 	node->realloc = realloc;
 	node->next = NULL;
-	if (zend_mm_observers == NULL) {
-		zend_mm_observers = node;
-		return zend_mm_observers;
+
+	// set bitflag for observers being around
+	heap->use_custom_heap |= ZEND_MM_CUSTOM_HEAP_OBSERVED;
+
+	if (heap->observers == NULL) {
+		heap->observers = node;
+		return node;
 	}
-	zend_mm_observer *current = zend_mm_observers;
+	zend_mm_observer *current = heap->observers;
 	while (current->next != NULL) {
 		current = current->next;
 	}
 	current->next = node;
+
 	return node;
 #else
 	return NULL;
 #endif
 }
 
 /*
- * This function will activate the global registered observers for the current
- * heap. After this has been run, observering functions will be called.
+ * Use this function to unregister your observer at any given time. In case your
+ * observer was the last one to be unregistered, this will also reset the
+ * `heap->use_custom_heap` bitflag.
  */
-void zend_mm_observers_startup(zend_mm_heap *heap)
+ZEND_API bool zend_mm_observer_unregister(zend_mm_heap *heap, zend_mm_observer *observer)
 {
 #if ZEND_MM_CUSTOM
-	if (heap == NULL) {
-		heap = AG(mm_heap);
-	}
-	zend_mm_observer *current_observer = zend_mm_observers;
-	if (current_observer == NULL) {
-		// no observers installed
-		return;
+	if (!heap && !(heap = AG(mm_heap))) {
+		return false;
 	}
-	// copy observers into heap
-	zend_mm_heap_observer *node = NULL;
-	zend_mm_heap_observer *current_heap_observer = NULL;
-	while (current_observer != NULL) {
-		node = pemalloc(sizeof(zend_mm_heap_observer), 1);
-		node->observer = current_observer;
-		node->enabled = true;
-		node->next = NULL;
-		if (current_heap_observer == NULL) {
-			heap->observers = node;
-		} else {
-			current_heap_observer->next = node;
+
+	zend_mm_observer *current = heap->observers, *prev = NULL;
+
+	if (current == observer) {
+		heap->observers = observer->next;
+		pefree(observer, 1);
+		if (!heap->observers)
+		{
+			// this was the one and only installed observer
+			heap->use_custom_heap &= ~ZEND_MM_CUSTOM_HEAP_OBSERVED;
 		}
-		current_heap_observer = node;
-		current_observer = current_observer->next;
+		return true;
 	}
-	// set bitflag for observers being around
-	heap->use_custom_heap |= ZEND_MM_CUSTOM_HEAP_OBSERVED;
+
+	while (current != NULL && current != observer) {
+		prev = current;
+		current = current->next;
+	}
+
+	// did not find observer or NULL was given
+	if (current == NULL)
+		return false;
```

Comment:
Nit: Add braces around the statement (single line if statements are error prone)